dali/python/nvidia/dali/external_source.py

```diff
@@ -222,9 +222,14 @@ def _is_generator_function(x):
     where __call__ is a generator function"""
     if inspect.isgeneratorfunction(x):
         return True
+    if isinstance(x, functools.partial):
+        return _is_generator_function(x.func)
+
     if x is None or inspect.isfunction(x) or inspect.ismethod(x):
         return False
     call = getattr(x, "__call__", None)
+    if call == x:
+        return False
```

Comment:
Without the lines 225-227 we got here. Then the functools wrapper ended up having `__call__` point to the enclosing object which triggered infinite recursion. This condition prevents this from happening for in other cases. Note that the worst thing that can happen here is that we miss a generator function - much less problematic than crashing on an otherwise good callback.

Reply:
What happens if we miss generator function here?

Reply:
Are there any other cases where call==x?

Reply:
> What happens if we miss generator function here?

It won't work as a generator function - so the particular use case is not handled


> Are there any other cases where call==x?

None so far, but I'd venture a guess that there might be something implemented in a similar way that `functools.partial` was.